src/pymor/algorithms/randrangefinder.py

```diff
@@ -0,0 +1,143 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2019 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+from scipy.sparse.linalg import eigsh, LinearOperator
+from scipy.special import erfinv
+
+from pymor.algorithms.gram_schmidt import gram_schmidt
+from pymor.core.defaults import defaults
+from pymor.operators.interfaces import OperatorInterface
+
+
+@defaults('tol', 'failure_tolerance', 'num_testvecs')
+def adaptive_rrf(A, source_product=None, range_product=None, tol=1e-4,
+                 failure_tolerance=1e-15, num_testvecs=20, lambda_min=None, iscomplex=False):
+    r"""Adaptive randomized range approximation of `A`.
+
+    This is an implementation of Algorithm 1 in [BS18]_.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `B` with the property
+
+    .. math::
+        \Vert A - P_{span(B)} A \Vert \leq tol
+
+    with a failure probability smaller than `failure_tolerance`, where the norm denotes the
+    operator norm. The inner product of the range of `A` is given by `range_product` and
+    the inner product of the source of `A` is given by `source_product`.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    source_product
+        Inner product |Operator| of the source of A.
+    range_product
+        Inner product |Operator| of the range of A.
+    tol
+        Error tolerance for the algorithm.
+    failure_tolerance
+        Maximum failure probability.
+    num_testvecs
+        Number of test vectors.
+    lambda_min
+        The smallest eigenvalue of source_product.
+        If `None`, the smallest eigenvalue is computed using scipy.
+    iscomplex
+        If `True`, the random vectors are chosen complex.
+
+    Returns
+    -------
+    B
+        |VectorArray| which contains the basis, whose span approximates the range of A.
+    """
+
+    assert source_product is None or isinstance(source_product, OperatorInterface)
+    assert range_product is None or isinstance(range_product, OperatorInterface)
+    assert isinstance(A, OperatorInterface)
+
+    B = A.range.empty()
+
+    R = A.source.random(num_testvecs, distribution='normal')
+    if iscomplex:
+        R += 1j*A.source.random(num_testvecs, distribution='normal')
+
+    if source_product is None:
+        lambda_min = 1
+    elif lambda_min is None:
+        def mv(v):
+            return source_product.apply(source_product.source.from_numpy(v)).to_numpy()
+
+        def mvinv(v):
+            return source_product.apply_inverse(source_product.range.from_numpy(v)).to_numpy()
```

Comment:
Is `adaptive_rrf` actually supposed to work if `source_product.space.from_numpy` is not implemented? Or is `from_numpy` always implemented for `source_products` in applications?

Reply:
> Or is from_numpy always implemented for source_products in applications?

No, this cannot be expected. In that case you would need to specify `lambda_min`. For the future, I would hope that at some point we reimplement something like ARPACK in pyMOR. Not sure how much effort would be required, however.